lib/internal/repl.js

```diff
@@ -162,62 +160,25 @@ function setupHistory(repl, historyPath, oldHistoryPath, ready) {
         }
       }
     }
-
-    fs.open(historyPath, 'w', onhandle);
-  }
-
-  function onhandle(err, hnd) {
-    if (err) {
-      return ready(err);
-    }
-    repl._historyHandle = hnd;
-    repl.on('line', online);
-
-    // reading the file data out erases it
-    repl.once('flushHistory', function() {
-      repl.resume();
-      ready(null, repl);
-    });
-    flushHistory();
-  }
-
-  // ------ history listeners ------
-  function online() {
-    repl._flushing = true;
-
-    if (timer) {
-      clearTimeout(timer);
-    }
-
-    timer = setTimeout(flushHistory, kDebounceHistoryMS);
+    // Flush history on close
+    repl.on('close', flushHistory);
+    repl.resume();
+    ready(null, repl);
   }
 
   function flushHistory() {
-    timer = null;
-    if (writing) {
-      pending = true;
-      return;
-    }
-    writing = true;
-    const historyData = repl.history.join(os.EOL);
-    fs.write(repl._historyHandle, historyData, 0, 'utf8', onwritten);
-  }
-
-  function onwritten(err, data) {
-    writing = false;
-    if (pending) {
-      pending = false;
-      online();
-    } else {
-      repl._flushing = Boolean(timer);
-      if (!repl._flushing) {
-        repl.emit('flushHistory');
-      }
+    const data = fs.readFileSync(historyPath, 'utf8');
+    const newLines = repl.history.length - initialHistoryLength;
+    let history = repl.history.slice(0, newLines);
+    if (data) {
+      history = history.concat(parseData(data)).slice(0, repl.historySize);
     }
+    const fd = fs.openSync(historyPath, 'a');
+    fs.writeSync(fd, history.join('\n'), 0, 'utf8');
```

Comment:
Right on all points. I have been using ^D to exit the REPL in my tests
which must account for the difference. I will address this.

On Tue, Jul 26, 2016 at 6:14 PM Anna Henningsen <notifications@github.com>
wrote:

> In lib/internal/repl.js
> <https://github.com/nodejs/node/pull/7005#discussion_r72366519>:
>
> >      }
> > +    const fd = fs.openSync(historyPath, 'a');
> > +    fs.writeSync(fd, history.join('\n'), 0, 'utf8');
>
> > rm ~/.node_repl_history
> $ ./node
> > 11
> > 22
> > 33
> > .exit
> $ ./node
> > 44
> > 55
> > 66
> > .exit
> $ cat ~/.node_repl_history.exit321.exit654.exit321
>
> Eek. Also, I personally consider it good practice to have all text files
> terminate with newlines. :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/7005/files/a74ca07852baafc5944c142d4c103c2067bb12cc#r72366519>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA-UKgrWsySmgECsaBZY2tlwkI1jV5Rks5qZrDzgaJpZM4In8hN>
> .
>